staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,616 @@
+	if qs.numRequestsEnqueued == 0 && qs.getRequestsExecuting() == 0 {
+		// fork an invocation of the EmptyHandler.
+		qs.counter.Add(1)
+		go func() {
+			eh.HandleEmpty()
+		}()
+	}
+
+	qs.emptyHandler = eh
+}
+
+// Wait uses the given hashValue as the source of entropy
+// as it shuffle-shards a request into a queue and waits for
+// a decision on what to do with that request.  If tryAnother==true
+//  at return then the QueueSet has become undesirable and the client
+// should try to find a different QueueSet to use; execute and
+// afterExecution are irrelevant in this case.  Otherwise, if execute
+//  then the client should execute  the request and then call
+// afterExecution().  Otherwise the client should not execute the
+//  request and afterExecution is irrelevant.
+func (qs *queueSetImpl) Wait(ctx context.Context, hashValue uint64) (tryAnother, execute bool, afterExecution func()) {
+	// TODO(aaron-prindle) verify locking is correct
+
+	qs.lockAndSyncTime()
+	// TODO(aaron-prindle) verify and test quiescent
+	// A call to Wait while the system is quiescent will be rebuffed by
+	// returning `tryAnother=true`.
+	if qs.emptyHandler != nil {
+		qs.lock.Unlock()
+		return true, false, func() {}
+	}
+
+	// ========================================================================
+	// Step 1:
+	// 1) Start with shuffle sharding, to pick a queue.
+	// 2) Reject old requests that have been waiting too long
+	// 3) Reject current request if there is not enough concurrency shares and
+	// we are at max queue length
+	// 4) If not rejected, create a packet and enqueue
+	req := qs.timeoutOldRequestsAndRejectOrEnqueue(hashValue)
+	// req == nil means that the request was rejected - no remaining
+	// concurrency shares and at max queue length already
+	if req == nil {
+		qs.lock.Unlock()
+		metrics.AddReject(qs.name, "queue-full")
+		return false, false, func() {}
+	}
+	// ========================================================================
+
+	// ------------------------------------------------------------------------
+	// Step 2:
+	// 1) The next step is to invoke the method that dequeues as much as possible.
+
+	// This method runs a loop, as long as there
+	// are non-empty queues and the number currently executing is less than the
+	// assured concurrency value.  The body of the loop uses the fair queuing
+	// technique to pick a queue, dequeue the request at the head of that
+	// queue, increment the count of the number executing, and send true to
+	// the request's channel.
+	qs.dequeueWithChannelAsMuchAsPossible()
+	// ------------------------------------------------------------------------
+	qs.lock.Unlock()
+
+	// ************************************************************************
+	// Step 3:
+	// After that method finishes its loop and returns, the final step in Wait
+	// is to `select` (wait) on a message from the enqueud packet's channel
+	// and return appropriately.
+	qs.counter.Add(-1)
```

Comment:
Why would we decrement this counter right now?

Reply:
Because this goroutine has no more work to du until some other goroutine does something to unblock this one.

Reply:
The corresponding increments are just before sending to the dequeue channel or closing the done channel (the latter we only need to get right in test code, in production the counter is a no-op).

Reply:
Oh, so negative numbers count blocked goroutines? I was expecting the increment to happen before the decrement.

Reply:
Or maybe this line is in error? We haven't yet read from the dequeue channel, that happens on the next few lines.